readthedocs/builds/models.py

```diff
@@ -377,14 +377,29 @@ def commit_name(self):
         return self.identifier
 
     def get_absolute_url(self):
-        """Get absolute url to the docs of the version."""
+        """
+        Get the absolute URL to the docs of the version.
+
+        If the version doesn't have a successfully uploaded build, then we return the project's
+        dashboard page.
+
+        Because documentation projects can be hosted on separate domains, this function ALWAYS
+        returns with a full "http(s)://<domain>/" prefix.
+        """
         if not self.built and not self.uploaded:
-            return reverse(
-                'project_version_detail',
-                kwargs={
-                    'project_slug': self.project.slug,
-                    'version_slug': self.slug,
-                },
+            # TODO: Stop assuming protocol based on settings.DEBUG
+            # (this pattern is also used in builds.tasks for sending emails)
+            protocol = "http" if settings.DEBUG else "https"
+            return "{}://{}{}".format(
+                protocol,
+                settings.PRODUCTION_DOMAIN,
+                reverse(
```

Comment:
It would already have broken if this was the case.

In your example, you are referring to `Build.get_absolute_url` and not `Version.get_absolute_url`.

Please see my PR description as it already mentions the Q&A'ing of this.

Reply:
It would already have broken because if the version has a successfully uploaded build, it returns `self.project.get_docs_url(...)` hence is supposed to include `https://<domain>`

Reply:
If I had done a refactor of how we used `Version.get_absolute_url`, then the change would have been far more intrusive.

Currently, it returns `get_docs_url`

https://github.com/readthedocs/readthedocs.org/blob/b255f7807fc7e01be785825d17f760433f5f7a19/readthedocs/projects/models.py#L544-L555

So `Version.get_absolute_url` was _already_ returning with scheme and domain, that's why the `target_url` was valid for callbacks with successful builds.

I'm not sure if I can explain this much better. Can you elaborate what you are proposing? To expand this change to refactor all calls to `Version.get_absolute_url` and for instance add a `Version.get_docs_url` ?

Reply:
Oh, OK. I see your point. I didn't see that `Version.get_absolute_url` was returning `Project.get_docs_url` which in the end returns a full URL. I think this behavior is _wrong_ and should be avoided.

However, as you said, this was already working like that so I don't expect you to change it in this PR 👍🏼 . We should do this refactor and follow standard patterns when possible, but in this case the damage was already there before this PR 🥲 -- In fact, now that you mentioned it, this is something that bitten myself in multiple opportunities in the past.

Anyways, we are not going to do this refactor here since the change will be huge. I'm fine moving forward.

Please, double-check if we are using `.get_absolute_url` in any other repository in a way that it doesn't expect the full URL before merging.

Reply:
> We should do this refactor and follow standard patterns when possible, but in this case the damage was already there before this PR smiling_face_with_tear

You are absolutely right about following the standards. Here, the root-cause was also NOT following the standards of what `get_absolute_url` should do. Actually, they were followed just fine on `Project.get_absolute_url` AFAICT. The reason that we have this bug to begin with is this.

I'm gonna open an issue about it and note that "this" happened. There's a chance that the changes aren't too scary.

Reply:
Filed here: https://github.com/readthedocs/readthedocs.org/issues/10371

Reply:
I've been looking through code on readthedocs-corporate and ext-theme and haven't found anything. I'm running tests on readthedocs-corporate with the changes in readthedocs.org to see that all is still good or if something needs to be adjusted.